ksqldb-execution/src/main/java/io/confluent/ksql/execution/interpreter/InterpretedExpressionFactory.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.interpreter;
+
+import io.confluent.ksql.execution.expression.tree.Expression;
+import io.confluent.ksql.execution.interpreter.terms.Term;
+import io.confluent.ksql.execution.util.ExpressionTypeManager;
+import io.confluent.ksql.function.FunctionRegistry;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+
+public final class InterpretedExpressionFactory {
+
+  private InterpretedExpressionFactory() { }
+
+  public static InterpretedExpression create(
+      final Expression expression,
+      final LogicalSchema schema,
+      final FunctionRegistry functionRegistry,
+      final KsqlConfig ksqlConfig
+  ) {
+    try {
+      final ExpressionTypeManager expressionTypeManager
+          = new ExpressionTypeManager(schema, functionRegistry);
+      final SqlType returnType = expressionTypeManager.getExpressionSqlType(expression);
+      if (returnType == null) {
+        throw new KsqlException("NULL expression not supported");
```

Comment:
when does the user see this error? what's the workaround? can we document that in the error message so that it's actionable?

Reply:
Heh, I think that's just when there's just a null and nothing else.  That's the only way type can be null.  I'll add some comments there.

Reply:
My personal rule of thumb is that, if we believe this should never be the case a.k.a. if it ever throws, it means a bug in our code --- like, type checking somehow missed it --- then instead of throwing a KsqlException we should throw it as a non-checked IllegalStateException. WDYT?